src/plugins/intel_gpu/src/runtime/ze/ze_memory.cpp

```diff
@@ -0,0 +1,268 @@
+    , memory(engine, new_layout, type, mem_tracker)
+    , _buffer(buffer)
+    , _host_buffer(engine->get_context(), engine->get_device()) {
+}
+
+gpu_usm::gpu_usm(ze_engine* engine, const layout& new_layout, const ze::UsmMemory& buffer, std::shared_ptr<MemoryTracker> mem_tracker)
+    : lockable_gpu_mem()
+    , memory(engine, new_layout, detect_allocation_type(engine, buffer), mem_tracker)
+    , _buffer(buffer)
+    , _host_buffer(engine->get_context(), engine->get_device()) {
+}
+
+gpu_usm::gpu_usm(ze_engine* engine, const layout& layout, allocation_type type)
+    : lockable_gpu_mem()
+    , memory(engine, layout, type, nullptr)
+    , _buffer(engine->get_context(), engine->get_device())
+    , _host_buffer(engine->get_context(), engine->get_device()) {
+    auto mem_ordinal = engine->get_device_info().device_memory_ordinal;
+    switch (get_allocation_type()) {
+    case allocation_type::usm_host:
+        _buffer.allocateHost(_bytes_count);
+        break;
+    case allocation_type::usm_shared:
+        _buffer.allocateShared(_bytes_count, mem_ordinal);
+        break;
+    case allocation_type::usm_device:
+        _buffer.allocateDevice(_bytes_count, mem_ordinal);
+        break;
+    default:
+        OPENVINO_THROW("[GPU] Unknown unified shared memory type!");
+    }
+
+    m_mem_tracker = std::make_shared<MemoryTracker>(engine, _buffer.get(), layout.bytes_count(), type);
+}
+
+void* gpu_usm::lock(const stream& stream, mem_lock_type type = mem_lock_type::read_write) {
+    std::lock_guard<std::mutex> locker(_mutex);
+    if (0 == _lock_count) {
+        auto& _ze_stream = downcast<const ze_stream>(stream);
+        if (get_allocation_type() == allocation_type::usm_device) {
+            if (type != mem_lock_type::read) {
+                throw std::runtime_error("Unable to lock allocation_type::usm_device with write lock_type.");
+            }
+            GPU_DEBUG_LOG << "Copy usm_device buffer to host buffer." << std::endl;
+            _host_buffer.allocateHost(_bytes_count);
+            OV_ZE_EXPECT(zeCommandListAppendMemoryCopy(_ze_stream.get_queue(),
+                                    _host_buffer.get(),
+                                    _buffer.get(),
+                                    _bytes_count,
+                                    nullptr,
+                                    0,
+                                    nullptr));
+            OV_ZE_EXPECT(zeCommandListHostSynchronize(_ze_stream.get_queue(), default_timeout));
+            _mapped_ptr = _host_buffer.get();
+        } else {
+            _mapped_ptr = _buffer.get();
+        }
+    }
+    _lock_count++;
+    return _mapped_ptr;
+}
+
+void gpu_usm::unlock(const stream& /* stream */) {
+    std::lock_guard<std::mutex> locker(_mutex);
+    _lock_count--;
+    if (0 == _lock_count) {
+        if (get_allocation_type() == allocation_type::usm_device) {
+            _host_buffer.freeMem();
+        }
+        _mapped_ptr = nullptr;
+    }
+}
```

Comment:
These code is very similar to the one in `ocl_memory.cpp`. Could you refactor it to reduce duplication?